cardano-api/src/Cardano/Api/Typed.hs

```diff
@@ -1647,88 +1650,96 @@ toJSONmOfnChecks keys required total
 
 instance FromJSON MultiSigScript where
-  parseJSON = Aeson.withObject "MultiSigScript" $ \obj -> return (parseMultiSigScript obj)
+  parseJSON = Aeson.withObject "MultiSigScript" $ \obj -> parseMultiSigScript obj
 
 
-parseMultiSigScript :: HMS.HashMap Text Value -> MultiSigScript
-parseMultiSigScript hms = case  all' hms <|> any' hms <|> atLeast hms of
-                            Just mss -> mss
-                            Nothing -> error "No multi-signature script found"
+parseMultiSigScript :: HMS.HashMap Text Value -> Aeson.Parser MultiSigScript
+parseMultiSigScript hms = do anyM <- any' hms
+                             allM <- all' hms
+                             atLeastM <- atLeast hms
+                             case anyM <|> allM <|> atLeastM of
+                               Just mss -> return mss
+                               Nothing -> fail "No multi-signature script found"
  where
-  convertValueToMultiSigScript :: Value -> Maybe MultiSigScript
+  convertValueToMultiSigScript :: Value -> Aeson.Parser (Maybe MultiSigScript)
   convertValueToMultiSigScript (Object valueHm) =
      payKeyHash' valueHm <|> any' valueHm  <|> all' valueHm <|> atLeast valueHm
-  convertValueToMultiSigScript _ = Nothing
-
-  payKeyHash' :: HMS.HashMap Text Value -> Maybe MultiSigScript
-  payKeyHash' hm = case HMS.lookup "type" hm of
-                    Just (String "requireSignature") ->
-                      case HMS.lookup "paymentKeyHash" hm of
-                        Just (String pkh) -> Just . RequireSignature $ convertToHash pkh
-                        Just val -> error $ "\"paymentKeyHash\" value is not a payment key hash. Value: " <> show val
-                        Nothing -> error "\"paymentKeyHash\" field is empty in your multi-signature script."
-                    _ -> Nothing
-
-  gatherMultiSigScripts :: Vector Value -> [MultiSigScript]
-  gatherMultiSigScripts = catMaybes . Vector.toList . Vector.map convertValueToMultiSigScript
-
-  any' :: HMS.HashMap Text Value -> Maybe MultiSigScript
-  any' hm = case HMS.lookup "type" hm of
-              Just (String "any") -> case HMS.lookup "scripts" hm of
-                                       Just (Array anyValue) ->
-                                         Just . RequireAnyOf $ gatherMultiSigScripts anyValue
-                                       _ -> Nothing
-              _ -> Nothing
-
-  all' :: HMS.HashMap Text Value -> Maybe MultiSigScript
-  all' hm = case HMS.lookup "type" hm of
-             Just (String "all") -> case HMS.lookup "scripts" hm of
-                                      Just (Array allVectorValue) ->
-                                        Just . RequireAllOf $ gatherMultiSigScripts allVectorValue
-                                      _ -> Nothing
-             _ -> Nothing
-
-  atLeast :: HMS.HashMap Text Value -> Maybe MultiSigScript
+  convertValueToMultiSigScript _ = return Nothing
```

Comment:
I think this is dangerous. It should succeed or fail, but not return a `Nothing` that gets discarded (via `catMaybes`). Otherwise this will hide parse failures. People could have scripts where one member of the all/any/atLeast gets ignored and that could be catastrophic. It should fail to parse in such cases.